gapis/api/vulkan/command_builder_utils.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vulkan
+
+import (
+	"context"
+
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/memory"
+)
+
+type commandBuilderUtil struct {
+	allocations   *allocationTracker
+	readMemories  []*api.AllocResult
+	writeMemories []*api.AllocResult
+}
+
+func newCommandBuilderUtil(allocations *allocationTracker) *commandBuilderUtil {
+	return &commandBuilderUtil{
+		allocations:   allocations,
+		readMemories:  make([]*api.AllocResult, 0),
+		writeMemories: make([]*api.AllocResult, 0),
+	}
+}
+
+func (util *commandBuilderUtil) allocReadData(ctx context.Context, g *api.GlobalState, v ...interface{}) api.AllocResult {
+	allocateResult := util.allocations.AllocDataOrPanic(ctx, v...)
+	util.readMemories = append(util.readMemories, &allocateResult)
+	rng, id := allocateResult.Data()
+	g.Memory.ApplicationPool().Write(rng.Base, memory.Resource(id, rng.Size))
+	return allocateResult
+}
+
+func (util *commandBuilderUtil) allocWriteData(ctx context.Context, v ...interface{}) api.AllocResult {
+	allocateResult := util.allocations.AllocDataOrPanic(ctx, v...)
+	util.writeMemories = append(util.writeMemories, &allocateResult)
+	return allocateResult
+}
+
+func (util *commandBuilderUtil) observeNewCommand(cmd api.Cmd) {
+	for i := range util.readMemories {
```

Comment:
Also, as I was read this file, my initial thoughts were "oh god this is n^2 complexity" with the number of commands processed, but I guess readMemories/writeMemories get emptied per command (or with some regularity), or the scope of this object is limited to a small number of commands mitigating the performance impact. Reading further into the review will show me which of these is true, but it would be nice if we could make this more obvious from the outside to someone unfamiliar with the code.

Reply:
Answering my own question above, I see this helper class is created pretty much per command being constructed so we don't have to worry about the n^2 problem above. I wonder how we could make the usage of this class more obvious to the reader. To be honest, the this class seems to only exist to tie together other parts of the gapis infrastructure which should be more tightly integrated to automatically keep data synchronised without the need for this little side-mechanism to keep data consistent when creating commands.